clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -6027,6 +6027,33 @@ static bool EvaluateBinaryTypeTrait(Sema &Self, TypeTrait BTT, const TypeSourceI
     return cast<CXXRecordDecl>(rhsRecord->getDecl())
       ->isDerivedFrom(cast<CXXRecordDecl>(lhsRecord->getDecl()));
   }
+  case BTT_IsVirtualBaseOf: {
+    const RecordType *BaseRecord = LhsT->getAs<RecordType>();
+    const RecordType *DerivedRecord = RhsT->getAs<RecordType>();
+
+    if (!BaseRecord || !DerivedRecord) {
+      DiagnoseVLAInCXXTypeTrait(Self, Lhs, tok::kw___is_virtual_base_of);
+      DiagnoseVLAInCXXTypeTrait(Self, Rhs, tok::kw___is_virtual_base_of);
+      return false;
+    }
+
+    if (BaseRecord->isUnionType() || DerivedRecord->isUnionType())
+      return false;
+
+    if (!BaseRecord->isStructureOrClassType() ||
+        !DerivedRecord->isStructureOrClassType())
+      return false;
```

Comment:
It could be an enum (enums can't have an inheritance relationship) - but this check is enough, we don't need the union check above. 
I think this is a mismatch between the standard - which treats union as classes, and clang - which treats everything as record.

Reply:
> It could be an enum

I thought a *`TagType`* was either a `RecordType` (`struct`/`class`/`union`) or an `EnumType` (`enum`). Can a `RecordType` really be an `enum`?

Reply:
IIRC ObjC interfaces are also `RecordType`s, and I don't think we want to bother with them.

Reply:
(and I don't think you can have a virtual objective c interface so there is nothing of ObjC to support here)

Reply:
`isStructureOrClassType` is defined as follows:
https://github.com/llvm/llvm-project/blob/0ee32c45730c94be1b7d5fa60a0e8dff5751d014/clang/lib/AST/Type.cpp#L657-L663

So (1) ObjC doesn't have to inherit from `RecordType` to get involved, and (2) the function has a misleading name.

Reply:
Hmm, `isInterface()` is defined on `TagDecl`, so it seems that a `RecordType` can still not be an ObjcInterface from what I can tell. Looks like `RD->isInterface()` check in here is just pointless.

Reply:
It’s also maybe worth noting that the line containing the check is from 2014, so things may have been different back then